pkg/storage/tsdb/block/block.go

```diff
@@ -163,16 +163,25 @@ func Upload(ctx context.Context, logger log.Logger, bkt objstore.Bucket, blockDi
 		return cleanUp(logger, bkt, id, err)
 	}
 
-	var oerr error
 	src := filepath.Join(blockDir, SparseIndexHeaderFilename)
 	dst := filepath.Join(id.String(), SparseIndexHeaderFilename)
-	if _, err := os.Stat(src); err == nil {
-		if err := objstore.UploadFile(ctx, logger, bkt, src, dst); err != nil {
-			// Don't call cleanUp. Uploading sparse index headers is best effort.
-			level.Warn(logger).Log("msg", "failed to upload sparse index headers", "block", id.String(), "err", err)
-			oerr = &UploadError{err, FileTypeSparseIndexHeader}
+	hasSparseHeaderInfo := false
+    for _, f := range meta.Thanos.Files {
+        if f.RelPath == SparseIndexHeaderFilename {
+            hasSparseHeaderInfo = true
+            break
+        }
+    }
+
+	// upload sparse header info if GatherFileStats returned an entry for it, otherwise skip
+	if hasSparseHeaderInfo{
+		if err:=objstore.UploadFile(ctx, logger, bkt, src, dst); err != nil {
+			return cleanUp(logger, bkt, id, &UploadError{err, FileTypeSparseIndexHeader})
 		}
 	}
+	else{
+		level.Debug(logger).Log("msg","sparse index header entry not found, skipping upload", "block", id.String())
+	}
```

Comment:
can we do this in a goroutine like the other files are done above? that way we can make use of the cleanUp invocation that already exists